src/target.c

```diff
@@ -108,61 +111,107 @@ stumpless_add_entry( struct stumpless_target *target,
   }
 
   clear_error(  );
+ 
+  if( stumpless_get_option( target, STUMPLESS_OPTION_PERROR ) == STUMPLESS_OPTION_PERROR ){
+	// setting is_log_formatted true concludes that STUMPLESS_OPTION_PERROR option is
+	// set, thus only checking this flag is enough for further checks
+	// NOTE : Calling the format_entry twice because if the buffer is for target who
+	// require unformatted entry would need not to call format_entry.
+	is_log_formatted = true;
+  	builder = format_entry( entry, target );
+  	if( !builder ) {
+    		return -1;
+  	}
+  	buffer = strbuilder_get_buffer( builder, &builder_length );
+  }
 
   // function targets are not formatted
-  if( target->type == STUMPLESS_FUNCTION_TARGET ) {
+  if( target->type == STUMPLESS_FUNCTION_TARGET ) {	
+    // Checking for the PERROR option
+    if( is_log_formatted ){
+    	error_result = stumpless_entry_to_errorstream( buffer, builder_length );    
+    	if(error_result == 0){
+	  return -1;
+	}
+    }
     return send_entry_to_function_target( target, entry );
   }
 
   // journald targets are not formatted
   if( target->type == STUMPLESS_JOURNALD_TARGET ) {
+    if( is_log_formatted ){
+    	error_result = stumpless_entry_to_errorstream( buffer, builder_length );
+    	if(error_result == 0)
+	  return -1;
+    }
     return config_send_entry_to_journald_target( target, entry );
   }
 
   // windows targets are not formatted in code
   // instead their formatting comes from message text files
   if( target->type == STUMPLESS_WINDOWS_EVENT_LOG_TARGET ) {
+     if( is_log_formatted ){
+    	error_result = stumpless_entry_to_errorstream( buffer, builder_length );
+     	if(error_result == 0)
+  	  return -1;
+     }
     return config_send_entry_to_wel_target( target->id, entry );
   }
 
-  builder = format_entry( entry, target );
-  if( !builder ) {
-    return -1;
-  }
-
-  buffer = strbuilder_get_buffer( builder, &builder_length );
+  // entry was not formatted before
+  if( is_log_formatted == false){
```

Comment:
Change this check to see if `builder` is `NULL` instead, to remove the need for a separate boolean flag.